src/linear_algebra/matrix/is_diag.lean

```diff
@@ -0,0 +1,189 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lu-Ming Zhang
+-/
+import linear_algebra.matrix.symmetric
+import linear_algebra.matrix.orthogonal
+import data.matrix.kronecker
+
+/-!
+# Diagonal matrices
+
+This file contains the definition and basic results about diagonal matrices.
+
+## Main results
+
+- `matrix.is_diag`: a proposition that states a given square matrix `A` is diagonal.
+
+## Tags
+
+diag, diagonal, matrix
+-/
+
+namespace matrix
+
+variables {α β R n m : Type*}
+
+open function
+open_locale matrix kronecker
+
+/-- `A.is_diag` means square matrix `A` is a diagonal matrix. -/
+def is_diag [has_zero α] (A : matrix n n α) : Prop := ∀ ⦃i j⦄, i ≠ j → A i j = 0
+
+@[simp] lemma is_diag_diagonal [has_zero α] [decidable_eq n] (d : n → α) :
+  (diagonal d).is_diag :=
+λ i j, matrix.diagonal_apply_ne
+
+/-- Diagonal matrices are generated by `matrix.diagonal`. -/
+lemma is_diag.exists_diagonal [has_zero α] [decidable_eq n] {A : matrix n n α} (h : A.is_diag) :
+  ∃ d, diagonal d = A :=
+begin
+  refine ⟨λ i, A i i, ext $ λ i j, _⟩,
+  obtain rfl | hij := decidable.eq_or_ne i j,
+  { rw diagonal_apply_eq },
+  { rw [diagonal_apply_ne hij, h hij] },
+end
+
+/-- `matrix.is_diag.exists_diagonal` as an iff. -/
+lemma is_diag_iff_exists_diagonal [has_zero α] [decidable_eq n] (A : matrix n n α) :
+  A.is_diag ↔ (∃ d, diagonal d = A) :=
+⟨is_diag.exists_diagonal, λ ⟨d, hd⟩, hd ▸ is_diag_diagonal d⟩
+
+/-- Every matrix indexed by a subsingleton is diagonal. -/
+lemma is_diag_of_subsingleton [has_zero α] [subsingleton n] (A : matrix n n α) : A.is_diag :=
+λ i j h, (h $ subsingleton.elim i j).elim
+
+/-- Every zero matrix is diagonal. -/
+@[simp] lemma is_diag_zero [has_zero α] : (0 : matrix n n α).is_diag :=
+λ i j h, rfl
+
+/-- Every identity matrix is diagonal. -/
+@[simp] lemma is_diag_one [decidable_eq n] [has_zero α] [has_one α] :
+  (1 : matrix n n α).is_diag :=
+λ i j, one_apply_ne
+
+lemma is_diag.map [has_zero α] [has_zero β]
+{A : matrix n n α} (ha : A.is_diag) {f : α → β} (hf : f 0 = 0) :
+  (A.map f).is_diag :=
+by { intros i j h, simp [ha h, hf] }
+
+lemma is_diag.neg [add_group α] {A : matrix n n α} (ha : A.is_diag) :
+  (-A).is_diag :=
+by { intros i j h, simp [ha h] }
+
+@[simp] lemma is_diag_neg_iff [add_group α] {A : matrix n n α} :
+  (-A).is_diag ↔ A.is_diag :=
+⟨ λ ha i j h, neg_eq_zero.1 (ha h), is_diag.neg ⟩
+
+lemma is_diag.add
+  [add_zero_class α] {A B : matrix n n α} (ha : A.is_diag) (hb : B.is_diag) :
+  (A + B).is_diag :=
+by { intros i j h, simp [ha h, hb h] }
+
+lemma is_diag.sub [add_group α]
+  {A B : matrix n n α} (ha : A.is_diag) (hb : B.is_diag) :
+  (A - B).is_diag :=
+by { intros i j h, simp [ha h, hb h] }
+
+lemma is_diag.smul [monoid R] [add_monoid α] [distrib_mul_action R α]
+  (k : R) {A : matrix n n α} (ha : A.is_diag) :
+  (k • A).is_diag :=
+by { intros i j h, simp [ha h] }
+
+@[simp] lemma is_diag_smul_one (n) [semiring α] [decidable_eq n] (k : α) :
```

Comment:
If the linter didn't like `is_diag_scalar` as a simp lemma, then I'd be inclined to just remove this lemma.

Reply:
that wasn't the problem. there was an unused argument. CI seems to be happy now

Reply:
ooh sorry, you are talking about an earlier commit.

Reply:
Yes, in an earlier commit `is_diag_scalar` was restated to be this lemma.